types/tx.go

```diff
@@ -28,12 +28,27 @@ type (
 	TxKey [TxKeySize]byte
 )
 
-// Hash computes the TMHASH hash of the wire encoded transaction.
+// Hash computes the TMHASH hash of the wire encoded transaction. It attempts to
+// unwrap the transaction if it is a MalleatedTx or a BlobTx.
 func (tx Tx) Hash() []byte {
+	if malleatedTx, isMalleated := UnwrapMalleatedTx(tx); isMalleated {
+		return tmhash.Sum(malleatedTx.Tx)
+	}
+	if blobTx, isBlobTx := UnmarshalBlobTx(tx); isBlobTx {
+		return tmhash.Sum(blobTx.Tx)
+	}
 	return tmhash.Sum(tx)
 }
 
+// Key returns the sha256 hash of the wire encoded transaction. It attempts to
+// unwrap the transaction if it is a BlobTx or a MalleatedTx.
 func (tx Tx) Key() TxKey {
+	if blobTx, isBlobTx := UnmarshalBlobTx(tx); isBlobTx {
+		return sha256.Sum256(blobTx.Tx)
+	}
+	if malleatedTx, isMalleated := UnwrapMalleatedTx(tx); isMalleated {
+		return sha256.Sum256(malleatedTx.Tx)
+	}
 	return sha256.Sum256(tx)
```

Comment:
[for reviewers]
I didn't abstract this, and instead optimized by putting the check that is most often used for each first. We mostly use tx.Key() in the mempool, so we should attempt to unmarshal blobs first, while we mostly use hash on wrapped (malleated) transactions, so we should do that first